builtin/providers/google/resource_storage_bucket.go

```diff
@@ -0,0 +1,154 @@
+package google
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"google.golang.org/api/storage/v1"
+)
+
+func resourceStorageBucket() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceStorageBucketCreate,
+		Read:   resourceStorageBucketRead,
+		Update: resourceStorageBucketUpdate,
+		Delete: resourceStorageBucketDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"predefined_acl": &schema.Schema{
+				Type:     schema.TypeString,
+				Default:  "private",
+				Optional: true,
+				ForceNew: true,
+			},
+			"location": &schema.Schema{
```

Comment:
Yeah this deserves more explanation :)

The value "US" comes from Google.  I suspect what is happening is you are telling Google to use a location of "" and the Go client libraries are interpreting that as "unset".

Since you're using ForceNew here, and the returned values from Google match what the user would provide, the distinction between Default and Computed is not so important.  However when we change this to be Update-able, it will cause a problem.

For completeness, here are the cases:

Optional Default:
Static default value, Terraform knows the default.  If not specified, use the default from the Terraform documenation and if the resource is not in that state, change the resource (possibly ForceNewing it).
Do not use .Set(), the correct value is already in the terraform state.
Example: http_health_check request_path

Optional:
As above, but default is zero / empty / etc as appropriate for the type.

Optional Computed:
If not specified, means user does not care.  Typically used when Terraform cannot pick a default, e.g. generated ip addresses that have to be globally unique.  Because the user does not care, Terraform will not change the resource if it sees the attribute has the wrong value.
Use .Set() to set the value from the response, if user did not provide a value.
Example: nat_ip in instance network_interface

Computed:
User cannot choose it (e.g. generated resource ids, self_link, etc).


If you use Optional Computed without ForceNew, then if the user deletes the attribute, Terraform will not return the value to the default.  A lot of resource attributes had this behavior originally, but users found it confusing in the case of simple default values (as this one).